### PR TITLE
FLW-1163 iOS: Extended networking & Chain registry approach

### PR DESCRIPTION
**Pull request by ERussel**
- complete chain registry implementation
- add integration test

### CODE REVIEW

File: fearless/Common/Services/ChainRegistry/ConnectionPool/ConnectionPool.swift [hunk 1]
```diff
@@ -2,7 +2,7 @@ import Foundation
 
 protocol ConnectionPoolProtocol {
     func setupConnection(for chain: ChainModel) throws -> ChainConnection
-    func getConnectionStates() throws -> [ConnectionPoolState]
+    func getConnetion(for chainId: ChainModel.Id) -> ChainConnection?
```
Comment (ef1rspb) [inline fearless/Common/Services/ChainRegistry/ConnectionPool/ConnectionPool.swift:5]: `getConnetion` -> `getConnection`

File: fearless/Common/Services/ChainRegistry/RuntimeFilesOperationFactory.swift [hunk 2]
```diff
@@ -1,24 +1,92 @@
...
+
+    /// Path to the directory where files are stored
     let directoryPath: String
 
+    /**
+     *  Creates instance a new instance for runtime types management.
+     *
+     *  - Parameters:
+     *      - repository: Engine that reads and writes files from filesystem;
+     *      - directory: Path to the directory where files are stored.
```
Comment (ef1rspb) [inline fearless/Common/Services/ChainRegistry/RuntimeFilesOperationFactory.swift:87]: `directoryPath`

File: fearless/Common/Services/ChainRegistry/RuntimeProviderPool/RuntimeProvider.swift [hunk 3]
```diff
@@ -195,6 +200,27 @@ extension RuntimeProvider: RuntimeProviderProtocol {
...
         resolveRequests()
     }
+
+    func fetchCoderFactoryOperation() -> BaseOperation<RuntimeCoderFactoryProtocol> {
+        ClosureOperation { [weak self] in
+            var fetchedFactory: RuntimeCoderFactoryProtocol?
+
+            let semaphore = DispatchSemaphore(value: 0)
+
+            self?.fetchCoderFactory(runCompletionIn: nil) { [weak semaphore] factory in
```
Comment (ef1rspb) [inline fearless/Common/Services/ChainRegistry/RuntimeProviderPool/RuntimeProvider.swift:210]: `fetchCoderFactory` stores reference to `semaphore`, but `semaphore` doesn't. I guess `[weak semaphore]` is redundant

File: fearless/Common/Services/ChainRegistry/SpecVersionSubscriptionFactory.swift [hunk 4]
```diff
@@ -1,16 +1,43 @@
...
 
+    /**
+     *  Creates new subscription factory
+     *
+     *  - Paramaters:
+     *      - runtimeSyncService: a sync service that is shared between
+     *      subscriptions created by the factory;
+     *      - logger: logger to provide info for debugging.
+     */
     init(runtimeSyncService: RuntimeSyncServiceProtocol, logger: LoggerProtocol) {
```
Comment (ef1rspb) [inline fearless/Common/Services/ChainRegistry/SpecVersionSubscriptionFactory.swift:41]: I think `logger` should be optional

**Review by ef1rspb — APPROVED** [review]

File: fearless/Common/Services/ChainRegistry/ChainRegistry.swift [hunk 5]
```diff
@@ -1,36 +1,212 @@
...
         self.connectionPool = connectionPool
         self.chainSyncService = chainSyncService
+        self.runtimeSyncService = runtimeSyncService
+        self.commonTypesSyncService = commonTypesSyncService
         self.chainProvider = chainProvider
         self.specVersionSubscriptionFactory = specVersionSubscriptionFactory
+        self.logger = logger
+
+        subscribeToChains()
+        syncUpServices()
```
[the remark this answers is not in the document]
Comment (ERussel) [reply fearless/Common/Services/ChainRegistry/ChainRegistry.swift:56]: Ok. Let's call `syncUp` manually

**Pull request merged**